src/Generator.cpp

```diff
@@ -1429,6 +1441,156 @@ Module GeneratorBase::build_module(const std::string &function_name,
+
+    const GeneratorParamInfo &pi = param_info();
+
+    // Even though propagate_adjoints() supports Funcs-of-Tuples just fine,
+    // we aren't going to support them here (yet); AFAICT, neither PyTorch nor
+    // TF support Tensors with Tuples-as-values, so we'd have to split the
+    // tuples up into separate Halide inputs and outputs anyway; since Generator
+    // doesn't support Tuple-valued Inputs at all, and Tuple-valued Outputs
+    // are quite rare, we're going to just fail up front, with the assumption
+    // that the coder will explicitly adapt their code as needed. (Note that
+    // support for Tupled outputs could be added with some effort, so if this
+    // is somehow deemed critical, go for it)
+    for (const auto *input : pi.inputs()) {
+        const size_t tuple_size = input->types_defined() ? input->types().size() : 1;
+        // Note: this should never happen
+        internal_assert(tuple_size == 1) << "Tuple Inputs are not yet supported by build_gradient_module()";
+    }
+    for (const auto *output : pi.outputs()) {
+        const size_t tuple_size = output->types_defined() ? output->types().size() : 1;
+        internal_assert(tuple_size == 1) << "Tuple Outputs are not yet supported by build_gradient_module";
+    }
+
+    std::vector<Argument> gradient_inputs;
+
+    // First: the original inputs. Note that scalar inputs remain scalar,
+    // rather being promoted into zero-dimensional buffers.
+    for (const auto *input : pi.inputs()) {
+        // There can be multiple Funcs/Parameters per input if the input is an Array
+        internal_assert(input->parameters_.size() == input->funcs_.size());
+        for (const auto &p : input->parameters_) {
+            gradient_inputs.push_back(to_argument(p, p.is_buffer() ? Expr() : input->get_def_expr()));
+            debug(DBG) << "    gradient copied input is: " << gradient_inputs.back().name << "\n";
+        }
+    }
+
+    // Next: add a grad-input for each *original* output; these will
+    // be the same shape as the output (so we should copy estimates from
+    // those outputs onto these estimates).
+    // - If an output is an Array, we'll have a separate input for each array element.
+
+    std::vector<ImageParam> d_output_imageparams;
+    for (const auto *output : pi.outputs()) {
+        for (size_t i = 0; i < output->funcs().size(); ++i) {
+            const Func &f = output->funcs()[i];
+            const std::string output_name = output->array_name(i);
+            // output_name is something like "funcname_i"
+            const std::string grad_in_name = replace_all(grad_input_pattern, "$OUT$", output_name);
+            // TODO(srj): does it make sense for gradient to be a non-float type?
+            // For now, assume it's always float32 (unless the output was float64).
+            const Type grad_in_type = output->type() == Float(64) ? Float(64) : Float(32);
```

Comment:
We also support Float(16) if that matters.

Reply:
done